Enable XUnit Analyzers

ViktorHofer opened the pull request:
Contributes to https://github.com/dotnet/corefx/issues/38677

I disabled the analyzer on five projects which need a bit more work:
- src\System.Dynamic.Runtime\tests\System.Dynamic.Runtime.Tests.csproj
- src\System.Linq.Expressions\tests\System.Linq.Expressions.Tests.csproj
- src\System.Runtime.Extensions\tests\System.Runtime.Extensions.Tests.csproj
... and two private.xml ones

sharwell commented on src/Microsoft.CSharp/tests/AssignmentTests.cs:
```diff
@@ -154,8 +160,8 @@ public void PropertyIncompatibleAssigment(object lhs, object rhs, bool overflow,
-        [Theory, MemberData(nameof(PropertyBadAssigments))]
-        public void PropertyIncompatibleConstantAssigment(object lhs, object rhs, bool _, bool invalidCast)
+        [Theory, MemberData(nameof(PropertyIncompatibleConstantAssigment_TestData))]
+        public void PropertyIncompatibleConstantAssigment(object lhs, object rhs, bool invalidCast)
```
😕 Which issue was reported here?

Ah, unused theory parameter. We typically solve this by assigning the parameter to a discard:

```suggestion
        public void PropertyIncompatibleConstantAssigment(object lhs, object rhs, bool overflow, bool invalidCast)
```

Then on the first line of the method:

```csharp
// Not used by this test
_ = overflow;
```

ViktorHofer replied:
I decided to use the data transform approach as it's the cleanest one according to the xunit analyzer rules but I'm open for suggestions. Do you think the discard one is cleaner?

sharwell replied:
It's entirely up to the project maintainers to decide. We chose discards because they work with both the XUnit rule and the FXCop rule for unused parameters:

https://github.com/dotnet/roslyn/issues/34441#issuecomment-479171267

ViktorHofer replied:
Thanks, I applied your suggestion and will outline it in the meta-issue.

stephentoub commented:
Most of these changes don't seem particularly valuable.  Do we really need all of the rules enabled?  The most important thing is ensuring that tests that should be run are run, which means all theories should have member/inline/etc. datas, and public parameterless void/Task methods should either be made private or be made [Fact]s/[Theory]s.

ViktorHofer commented:
It's fairly easy to opt-out of analyzer rules. This can be applied per assembly or for multiple assemblies / repo wide.

> The most important thing is ensuring that tests that should be run are run, which means all theories should have member/inline/etc. datas, and public parameterless void/Task methods should either be made private or be made [Fact]s/[Theory]s.

That's definitely something to watch out for but IMO not the only reason to enable the xunit analyzers. I can come up with a list of rules that might not offer much value.

stephentoub commented:
> but IMO not the only reason to enable the xunit analyzers

Which others do you view as critically important?

I would rather we enable the "this is a test that should run but isn't" rules for all of corefx (I expect there are only a few violations, which we should be able to fix quickly), rather than delay rolling it out, gated on addressing lots of minor quibbles.

ViktorHofer commented:
> Which others do you view as critically important

I don't think this entire effort is critically important. Otherwise we would have a tracking issue that is set to 3.0. Therefore I'm not trying to find a quick solution here and I tend towards a black-listing approach.

I believe most of the errors make sense and should be enabled (https://xunit.net/xunit.analyzers/rules/):

- https://xunit.net/xunit.analyzers/rules/xUnit1000
- https://xunit.net/xunit.analyzers/rules/xUnit1001
- https://xunit.net/xunit.analyzers/rules/xUnit1002
- https://xunit.net/xunit.analyzers/rules/xUnit1003
- https://xunit.net/xunit.analyzers/rules/xUnit1005
- https://xunit.net/xunit.analyzers/rules/xUnit1006
- https://xunit.net/xunit.analyzers/rules/xUnit1008
- https://xunit.net/xunit.analyzers/rules/xUnit1009
- https://xunit.net/xunit.analyzers/rules/xUnit1010
- https://xunit.net/xunit.analyzers/rules/xUnit1011
- https://xunit.net/xunit.analyzers/rules/xUnit1013
- https://xunit.net/xunit.analyzers/rules/xUnit2016
- https://xunit.net/xunit.analyzers/rules/xUnit2014

Also I'm asking for uniformity here. Do we already have a ruleset defined somewhere for the dotnet organization that is considered as useful / do we have an exclusion list somewhere? It sounds like corefx is not the first repository that enables xunit analyzers.

sharwell commented:
💡 xUnit1004 (Test methods should not be skipped) is handy to keep at severity Hidden. Hidden severity eliminates the message from the error list but keeps the ability to use the diagnostic's code fix to remove the Skip attribute. Setting it to None disables the diagnostic entirely so the code fix will not be available to use as a refactoring.

stephentoub commented:
> I don't think this entire effort is critically important.

There's a significant difference between:
a) Tests someone wrote that should be running but aren't.
b) Style quibbles.

I do think (a) is pretty darn important.  It means things we expected to be tested aren't, and until we know what they are, we don't know whether they're critical or not.

Out of your list, to me that means we should just blanket enable:

- xUnit1000
- xUnit1003
- xUnit1013

Many of the others are about catching at compile time what will otherwise be caught at runtime, e.g. xUnit1001 catches cases where a [Fact] has parameters, and that will today cause runtime failures.  Would it be nice to catch at compile time instead?  Yeah, it'd be nice.  But it's not worth holding up the above for it. (That said, we should have 0 violations of these, exactly because they're checked at runtime.)

I would like to do this rather than doing it piecemeal, as this PR starts to do.  If you don't want to do it, I'll do it.

stephentoub commented on src/Microsoft.CSharp/tests/Microsoft.CSharp.Tests.csproj:
```diff
@@ -2,6 +2,7 @@
   <PropertyGroup>
     <ProjectGuid>{82B54697-0251-47A1-8546-FC507D0F3B08}</ProjectGuid>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <EnableXUnitAnalyzer>true</EnableXUnitAnalyzer>
```
Even with my top-level comments, this PR LGTM.  The primary thing I'm concerned about is this one line.  We should be doing this at the repo level at a minimum, and since you say that would cause a significant amount of work to achieve due to lots of warnings across the repo, scoping it to just the critical set of warnings about tests that should be running but aren't; there should be relatively few of those, and we should be able to enable them in one fell swoop.  If it turns out there are way more than "relatively few", that then just highlights we have a bigger issue than we thought and it's worth addressing even more.

ViktorHofer replied:
I get your point. I'll do the necessary to enable these rules and will provide an extensibility point so that assemblies can go the extra mile and enable more rules.

stephentoub replied:
Thank you! If you'd prefer I do it, just let me know.

ViktorHofer replied:
Everything's good. I'll do it :)

bartonjs replied:
Since this is on at the repo level now, should this line be deleted?

ViktorHofer commented:
@stephentoub as discussed I updated the PR and addressed the most important rules across the repo. I expect test failures as I enabled a lot of tests.

stephentoub commented on src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/RSA/KeyGeneration.cs:
```diff
@@ -21,10 +21,10 @@ public static void GenerateSecondMinKey()
+        // public static void GenerateMaxKey()
+        // {
+        //     GenerateKey(rsa => GetMax(rsa.LegalKeySizes));
+        // }
```
It'd be better to either a) just delete this, or b) make it a [ConditionalFact] to only run in some particular scenarios.
cc: @bartonjs

bartonjs replied:
Mainly it's there to answer the question "why is only one bound checked?".  It'd be something we'd only enable after changing the max size on a provider, then let it do its half-day execution, then turn it off again.

So, while my preference is to leave it exactly as it is, I understand that's not an option... so either Fact(Skip) or delete is fine.  (Not ConditionalFact, it doesn't need to show up as a "hey, there's a test I didn't run.)

stephentoub replied:
> Not ConditionalFact, it doesn't need to show up as a "hey, there's a test I didn't run

Doesn't Skip show up like that, too? Maybe I'm misremembering.

bartonjs replied:
Well, at least one of us is. :smile:.  If they're both printed as skipped then I'm okay with some sort of `[ConditionalFact(nameof(WillingToWait8HoursForOneTest))]`

stephentoub replied:
If it actually takes 8hrs to run, is anyone ever going to run it? Maybe just delete it.

sharwell commented on src/CodeAnalysis.ruleset:
```diff
@@ -38,4 +38,47 @@
+    <Rule Id="xUnit1010" Action="None" />
+    <Rule Id="xUnit1011" Action="None" />
+    <Rule Id="xUnit1012" Action="None" />
+    <Rule Id="xUnit1013" Action="Error" />
```
💡 Use warnings-as-errors to escalate warnings to errors.

```suggestion
    <Rule Id="xUnit1013" Action="Warning" />
```

ViktorHofer replied:
You mean basically for all rules? As we enabled warningsAsErrors=true it doesn't matter much.

ViktorHofer replied:
OK got what you said. applied, thanks.

sharwell commented on src/Common/tests/System/Security/Cryptography/AlgorithmImplementations/RSA/KeyGeneration.cs:
```diff
@@ -21,10 +21,10 @@ public static void GenerateSecondMinKey()
+        // public static void GenerateMaxKey()
+        // {
+        //     GenerateKey(rsa => GetMax(rsa.LegalKeySizes));
+        // }
```
💡 Either delete the test, or use `[Fact(Skip = "explanation")]`

sharwell commented on src/Common/tests/System/Xml/XPath/TestsWithEncoding/MiscWithEncodings.cs:
```diff
@@ -16,7 +16,7 @@ namespace XPathTests.FunctionalTests.Location.Paths
     public static partial class MiscWithEncodingTests
     {
-        public static void TestInitialize()
+        private static void TestInitialize()
```
💭 This works, but normally I would expect to see this code in the constructor for the test class.

sharwell commented on src/System.Collections.Immutable/tests/ImmutableSortedDictionaryTest.cs:
```diff
@@ -425,8 +425,8 @@ public static void TestDebuggerAttributes_Null()
-        ////[Fact] // not really a functional test -- but very useful to enable when collecting perf traces.
-        public void EnumerationPerformance()
+        // [Fact] // not really a functional test -- but very useful to enable when collecting perf traces.
+        private void EnumerationPerformance()
```
💡 To distinguish this case from other helpers, it would be preferable to use `[Fact(Skip = "explanation")]` rather than mark it private.

sharwell commented on src/System.Collections.Immutable/tests/ImmutableSortedDictionaryTest.cs:
```diff
@@ -450,8 +450,8 @@ public void EnumerationPerformance()
-        ////[Fact] // not really a functional test -- but very useful to enable when collecting perf traces.
-        public void EnumerationPerformance_Empty()
+        // [Fact] // not really a functional test -- but very useful to enable when collecting perf traces.
+        private void EnumerationPerformance_Empty()
```
💡 To distinguish this case from other helpers, it would be preferable to use `[Fact(Skip = "explanation")]` rather than mark it private.

sharwell commented on src/System.Collections.NonGeneric/tests/QueueTests.cs:
```diff
@@ -956,7 +956,7 @@ public static void Synchronized_NullQueue_ThrowsArgumentNullException()
         }
 
-        public void StartEnqueueThread()
+        private void StartEnqueueThread()
```
💡 Consider making this a local function since it's only used by one test

stephentoub commented on src/System.CodeDom/tests/System/CodeDom/TempFileCollectionTests.cs:
```diff
@@ -32,6 +32,8 @@ protected override void AddToCollection(ICollection collection, int numberOfItem
         protected override Type ICollection_NonGeneric_CopyTo_ArrayOfIncorrectReferenceType_ThrowType => typeof(InvalidCastException);
 
+        [Theory]
+        [MemberData(nameof(ValidCollectionSizes))]
```
This seems wrong to me.  The base test has such attributes on it, doesn't it?  I'd expect those to be picked up by xunit, and if they're not, that seems like an xunit bug.  And an xunit analyzer bug that it's flagging this.

ViktorHofer replied:
Yes, this is an xunit analyzer bug. They are picked up correctly. I'll file an issue an meanwhile disable the warning locally here.

ViktorHofer replied:
https://github.com/xunit/xunit/issues/1973

sharwell commented on src/System.Collections/tests/Generic/LinkedList/LinkedList.Generic.Tests.Constructor.cs:
```diff
@@ -98,7 +98,6 @@ internal LinkedList_T_Tests(LinkedList<T> collection, T[] expectedItems)
         /// <summary>
         /// Tests the initial items in the list.
         /// </summary>
-        [Fact]
```
😕 Why was this removed?

ViktorHofer replied:
Thsi is a helper method inside an internal class which is invoked from test methods in another class.

stephentoub commented on src/System.Collections.Immutable/tests/ImmutableSortedDictionaryTest.cs:
```diff
@@ -450,8 +450,8 @@ public void EnumerationPerformance()
 
-        ////[Fact] // not really a functional test -- but very useful to enable when collecting perf traces.
-        public void EnumerationPerformance_Empty()
+        // [Fact] // not really a functional test -- but very useful to enable when collecting perf traces.
```
Then we should make these ConditionalFacts on, for example, some environment variable being set.

stephentoub commented on src/System.Console/tests/WindowAndCursorProps.cs:
```diff
@@ -270,12 +270,14 @@ public static void Title_Set_Windows(int lengthOfTitle)
         }, lengthOfTitle.ToString()).Dispose();
     }
 
+    [Fact]
```
Based on the test name, shouldn't this just be for UAP?  Same below.

stephentoub commented on src/System.Diagnostics.Process/tests/ProcessTests.cs:
```diff
@@ -1343,14 +1343,14 @@ public void StandardInput_GetNotRedirected_ThrowsInvalidOperationException()
+        //         Console.WriteLine("{0} : \"{1}\" (Threads: {2})", p.Id, p.ProcessName, p.Threads.Count);
+        //         p.Dispose();
+        //     }
+        // }
```
Just delete this.

sharwell replied:
💡 It would be preferable to use `[Fact(Skip = "explanation")]` here

stephentoub commented on src/System.Dynamic.Runtime/tests/System.Dynamic.Runtime.Tests.csproj:
```diff
@@ -3,6 +3,7 @@
     <ProjectGuid>{0BFD6D9F-DF9E-4B17-8ED4-29437AE5B04A}</ProjectGuid>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <NoWarn>67,168,219,414,162,184,458,464,78,169,114,693,108,1981,649,109,1066,3021,3026,3002,3014,3022,660,661,429</NoWarn>
+    <EnableXUnitAnalyzer>false</EnableXUnitAnalyzer>
```
Why?

stephentoub commented on src/System.Linq.Expressions/tests/System.Linq.Expressions.Tests.csproj:
```diff
@@ -2,14 +2,13 @@
     <DefineConstants Condition=" '$(IsInterpreting)' != 'true' ">$(DefineConstants);FEATURE_COMPILE</DefineConstants>
     <DefineConstants Condition=" '$(FeatureInterpret)' == 'true' ">$(DefineConstants);FEATURE_INTERPRET</DefineConstants>
     <IncludeDefaultReferences>false</IncludeDefaultReferences>
+    <EnableXUnitAnalyzer>false</EnableXUnitAnalyzer>
```
Why?

stephentoub commented on src/System.Private.Xml.Linq/tests/Schema/ExtensionTests.cs:
```diff
@@ -86,28 +86,28 @@ public SchemaExtensionsTests()
+		// 	reader = XmlReader.Create(xmlDocumentPath, settings);
+		// 	xmlDocument = XDocument.Load(reader);
+		// 	validationSucceded = false;
+		// }
```
If it's not being used, just delete it.

sharwell replied:
❔ Should this just be deleted?

stephentoub commented on src/System.Private.Xml.Linq/tests/misc/RegressionTests.cs:
```diff
@@ -37,17 +37,17 @@ public void CannotParseDTD()
+        //         TestLog.Compare(x.Name.LocalName, s, s);
+        //         x = x.FirstNode as XElement;
+        //     }
+        // }
```
Why is this commented out rather than being made into a [Fact]?

sharwell replied:
💭 I can't tell if this is supposed to be a test or if it could maybe be deleted

stephentoub commented on src/System.Private.Xml.Linq/tests/xNodeBuilder/System.Xml.Linq.xNodeBuilder.Tests.csproj:
```diff
@@ -1,6 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <ProjectGuid>{5D4FB9ED-C3AC-4EFA-9FEE-619ED4B4B92D}</ProjectGuid>
+    <EnableXUnitAnalyzer>false</EnableXUnitAnalyzer>
```
Why?  I'll stop asking about these.

stephentoub commented on src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs:
```diff
@@ -59,20 +59,20 @@ public void ValidateBasicObjectsRoundtrip(object obj, FormatterAssemblyStyle ass
+        //     Console.WriteLine($"{numberOfBlobs} existing blobs" +
+        //         $"{Environment.NewLine}{numberOfFoundBlobs} found blobs with regex search" +
+        //         $"{Environment.NewLine}{numberOfUpdatedBlobs} updated blobs with regex replace");
+        // }
```
I don't like commenting these out.  I'd rather see it made into a [ConditionalFact] that checks for a particular environment variable.  Then to update the blobs, all one needs to do is set that environment variable and run the tests.

ViktorHofer replied:
This helper method doesn't work anymore, therefore I'm deleting it.

stephentoub commented on src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTests.cs:
```diff
@@ -443,37 +443,37 @@ public void ObjectReference_RealObjectSerialized()
+        //     catch (TargetInvocationException) { }
+        //     catch (ArgumentException) { }
+        //     catch (FileLoadException) { }
+        // }
```
Ditto

stephentoub commented on src/System.Runtime.Serialization.Xml/tests/DataContractSerializer.cs:
```diff
@@ -2421,13 +2421,13 @@ public static void DCS_GenericICollectionOfBoolean()
+    //     var deserialized = DataContractSerializerHelper.SerializeAndDeserialize(value, @"<ArrayOfdateTime xmlns=""http://schemas.microsoft.com/2003/10/Serialization/Arrays"" xmlns:i=""http://www.w3.org/2001/XMLSchema-instance""><dateTime>2000-01-02T03:04:05-08:00</dateTime><dateTime>2011-02-03T04:05:06-08:00</dateTime></ArrayOfdateTime>");
+    //     Assert.StrictEqual(value.Count, deserialized.Count);
+    //     Assert.StrictEqual(true, Enumerable.SequenceEqual(value, deserialized));
+    // }
```
Delete.

stephentoub commented on src/System.Runtime/tests/System/EnumTests.cs:
```diff
@@ -224,6 +224,7 @@ public static void Parse_Invalid(Type enumType, string value, bool ignoreCase, T
         }
 
+#pragma warning disable xUnit1013 // Public method should be marked as test
         public static void Parse_Generic_Invalid<T>(Type enumType, string value, bool ignoreCase, Type exceptionType) where T : struct
```
It shouldn't need to be public.  You can instead just change the GetMethod call above to using BingingFlags.Static | BindingFlags.NonPublic.

stephentoub commented on src/System.Runtime/tests/System/EnumTests.netcoreapp.cs:
```diff
@@ -39,6 +39,7 @@ public static void Parse_Invalid_NetCoreApp11(Type enumType, string value, bool
         }
 
+#pragma warning disable xUnit1013 // Public method should be marked as test
         public static void Parse_Generic_Invalid_NetCoreApp11<T>(Type enumType, string value, bool ignoreCase, Type exceptionType) where T : struct
```
Ditto

stephentoub commented on src/System.Security.Cryptography.Csp/tests/CspParametersTests.cs:
```diff
@@ -46,8 +46,7 @@ public static void KeyPassword_SetGet()
 
-        //Manual test - requires Smart Card - read instructions
-        //[Theory]
+        [Theory(Skip = "Manual test - requires Smart Card - read instructions")]
```
Ditto on env var or something similar.
cc: @bartonjs

bartonjs replied:
Presumably `container` could get its value from an env var (TEST_CSP_SMARTCARD_CONTAINER?) and then a condition check if there's a value for that container.

But, since this is a very manual test (it'll pop UI), I'm good with Fact(Skip)

stephentoub commented on src/System.Transactions.Local/tests/NonMsdtcPromoterTests.cs:
```diff
@@ -78,7 +78,9 @@ private static void CompleteDependentCloneThread(object stateObject)
 
+#pragma warning disable xUnit1013 // Public method should be marked as test
         public static void Promote(string testCaseDescription, byte[] promotedTokenToCompare, Transaction txToPromote = null)
+#pragma warning restore xUnit1013 // Public method should be marked as test
```
Why does it need to be public?

stephentoub commented:
(Please make sure to squash this all when merging.)

bartonjs commented on src/Directory.Build.props:
```diff
@@ -6,8 +6,8 @@
     <DisableImplicitConfigurationDefines>true</DisableImplicitConfigurationDefines>
-
     <CodeAnalysisRuleset>$(MSBuildThisFileDirectory)CodeAnalysis.ruleset</CodeAnalysisRuleset>
+    <EnableXUnitAnalyzer>true</EnableXUnitAnalyzer>
```
Is it a problem (noticable build time increase, etc) to have this for the src projects as well as test? (Should this line be in a different props file or a propertygroup... wherever we have stuff that is different for tests (collectively)?)

ViktorHofer replied:
No this property is simply ignored on source projects, we don't have a separate place for source vs test properties.

bartonjs commented on src/System.Runtime.Loader/tests/DefaultContext/DefaultLoadContextTest.cs:
```diff
@@ -183,7 +183,7 @@ public static void LoadNonExistentInDefaultContext()
         }
 
-        public void DefaultContextFallback()
+        private void DefaultContextFallback()
```
Why are these being marked private instead of `[Fact]`?

ViktorHofer replied:
Because these are helper methods which are invoked in the test method above.

bartonjs commented on src/System.Text.Encoding.CodePages/tests/EncodingCodePages.cs:
```diff
@@ -633,7 +633,7 @@ public static void TestEncodingDisplayNames(int codePage, string webName, string
         // This test is run as part of the default mappings test, since it modifies global state which that test
         // depends on.
-        public static void TestRegister1252()
+        private static void TestRegister1252()
```
I wonder if the comment meant "is not run".  But I'm not sure what heuristic was applied for private vs Fact(Skip) vs comment out vs delete.

ViktorHofer replied:
The comment means that is invoked by a method which is marked as fact as the parent method depends on global state manipulation.

bartonjs commented on src/System.Text.Json/tests/Serialization/Value.ReadTests.GenericCollections.cs:
```diff
@@ -729,6 +729,7 @@ public static void ReadPrimitiveKeyValuePairFail()
             Assert.Throws<JsonException>(() => JsonSerializer.Parse<KeyValuePair<string, int>>(@"{""Key"": ""Key"", ""Val"": 123"));
         }
 
+       [Fact]
```
Nit: Misaligned whitespace :smile:

bartonjs commented on src/System.Transactions.Local/tests/NonMsdtcPromoterTests.cs:
```diff
@@ -1585,7 +1587,7 @@ public static void TestCase_TransactionTimeout(bool promote)
         }
 
-        public static void TestCase_GetStatusAndDistributedId()
+        private static void TestCase_GetStatusAndDistributedId()
```
I see that this one, which looks just like a `[Fact]` to me, is called by a `[Fact]` (which does nothing but call this).  I'll assume that is true for the rest of the void-no-arg public to private conversions in this file.

ViktorHofer replied:
Yeah, this doesn't look right but I'm not changing this as part of this PR.

bartonjs commented on src/System.ValueTuple/tests/TupleTests.cs:
```diff
@@ -27,15 +27,15 @@ public void Test()
         }
 
-        public static void OneTuples()
+        private static void OneTuples()
```
I'm not sure I understand a) why ValueTuple.Tests is testing Tuple, or b) why this file uses one driver method to call all the tests and the ValueTupleTests one uses individual facts.

I propose that deleting `public void Test()` and marking each of these as `[Fact]` is a better change.

ViktorHofer commented:
@sharwell @stephentoub @bartonjs thanks for the code review. I applied all suggestions and am now looking at the test failures.

ViktorHofer commented:
@stephentoub can you please look at the failing re-enabled System.Console tests? https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F38676~2Fmerge/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190701.40/workItem/System.Console.Tests

Thanks!

ViktorHofer commented:
This is ready to be **squashed** into master. I opened https://github.com/dotnet/corefx/issues/39233 for 
follow-up work.

stephentoub commented:
Thanks, @ViktorHofer.

ViktorHofer commented:
No need to thank me, we all care about test quality. I'm glad that we know have static analysis to capture wrong annotations.
